File: app/src/androidTest/java/com/thewizrd/simpleweather/UnitTests.java
package com.thewizrd.simpleweather;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.util.Log;

import androidx.test.core.app.ApplicationProvider;
import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.google.android.gms.tasks.Tasks;
import com.google.common.base.Stopwatch;
import com.jakewharton.threetenabp.AndroidThreeTen;
import com.thewizrd.shared_resources.AppState;
import com.thewizrd.shared_resources.ApplicationLib;
import com.thewizrd.shared_resources.SimpleLibrary;
import com.thewizrd.shared_resources.controls.LocationQueryViewModel;
import com.thewizrd.shared_resources.locationdata.LocationData;
import com.thewizrd.shared_resources.tasks.AsyncTask;
import com.thewizrd.shared_resources.tasks.CallableEx;
import com.thewizrd.shared_resources.tzdb.TimeZoneProvider;
import com.thewizrd.shared_resources.utils.JSONParser;
import com.thewizrd.shared_resources.utils.Logger;
import com.thewizrd.shared_resources.utils.Settings;
import com.thewizrd.shared_resources.utils.StringUtils;
import com.thewizrd.shared_resources.utils.WeatherException;
import com.thewizrd.shared_resources.utils.WeatherUtils;
import com.thewizrd.shared_resources.utils.here.HEREOAuthUtils;
import com.thewizrd.shared_resources.weatherdata.Astronomy;
import com.thewizrd.shared_resources.weatherdata.Weather;
import com.thewizrd.shared_resources.weatherdata.WeatherAPI;
import com.thewizrd.shared_resources.weatherdata.WeatherAlert;
import com.thewizrd.shared_resources.weatherdata.WeatherManager;
import com.thewizrd.shared_resources.weatherdata.WeatherProviderImpl;
import com.thewizrd.shared_resources.weatherdata.images.ImageDatabase;
import com.thewizrd.shared_resources.weatherdata.nws.SolCalcAstroProvider;
import com.thewizrd.shared_resources.weatherdata.nws.alerts.NWSAlertProvider;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.threeten.bp.LocalDateTime;
import org.threeten.bp.ZonedDateTime;
import org.threeten.bp.format.DateTimeFormatter;

import java.io.IOException;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;

@RunWith(AndroidJUnit4.class)
public class UnitTests {
    @Before
    public void init() {
        // Context of the app under test.
        final Context appContext = ApplicationProvider.getApplicationContext();

        ApplicationLib app = new ApplicationLib() {
            @Override
            public Context getAppContext() {
                return appContext.getApplicationContext();
            }

            @Override
            public SharedPreferences getPreferences() {
                return PreferenceManager.getDefaultSharedPreferences(getAppContext());
            }

            @Override
            public SharedPreferences.OnSharedPreferenceChangeListener getSharedPreferenceListener() {
                return null;
            }

            @Override
            public AppState getAppState() {
                return null;
            }

            @Override
            public boolean isPhone() {
                return true;
            }
        };

        SimpleLibrary.init(app);
        AndroidThreeTen.init(appContext);

        // Start logger
        Logger.init(appContext);
        Settings.loadIfNeeded();
    }

    private Weather getWeather(WeatherProviderImpl providerImpl) throws WeatherException {
        LocationQueryViewModel location = providerImpl.getLocation(new WeatherUtils.Coordinate(47.6721646, -122.1706614));
        LocationData locData = new LocationData(location);
        return providerImpl.getWeather(locData);
    }

    @Test
    public void getHEREWeather() throws WeatherException {
        WeatherProviderImpl provider = WeatherManager.getProvider(WeatherAPI.HERE);
        Weather weather = getWeather(provider);
        Assert.assertTrue(weather != null && weather.isValid());
    }

    @Test
    public void getYahooWeather() throws WeatherException {
        WeatherProviderImpl provider = WeatherManager.getProvider(WeatherAPI.YAHOO);
        Weather weather = getWeather(provider);
        Assert.assertTrue(weather != null && weather.isValid());
    }

    @Test
    public void getMetNoWeather() throws WeatherException {
        WeatherProviderImpl provider = WeatherManager.getProvider(WeatherAPI.METNO);
        Weather weather = getWeather(provider);
        Assert.assertTrue(weather != null && weather.isValid());
    }

    @Test
    public void getNWSAlerts() throws WeatherException {
        LocationQueryViewModel location = WeatherManager.getProvider(WeatherAPI.OPENWEATHERMAP)
                .getLocation(new WeatherUtils.Coordinate(47.6721646, -122.1706614));
        LocationData locData = new LocationData(location);
        List<WeatherAlert> alerts = new NWSAlertProvider().getAlerts(locData);
        Assert.assertNotNull(alerts);
    }

    @Test
    public void getNWSWeather() throws WeatherException {
        WeatherProviderImpl provider = WeatherManager.getProvider(WeatherAPI.NWS);
        Weather weather = getWeather(provider);
        Assert.assertTrue(weather != null && weather.isValid());
    }

    @Test
    public void getOWMWeather() throws WeatherException {
        WeatherProviderImpl provider = WeatherManager.getProvider(WeatherAPI.OPENWEATHERMAP);
        Weather weather = getWeather(provider);
        Assert.assertTrue(weather != null && weather.isValid());
    }

    @Test
    public void getHEREOAuthToken() {
        String token = AsyncTask.await(new Callable<String>() {
            @Override
            public String call() {
                return HEREOAuthUtils.getBearerToken(true);
            }
        });
        Assert.assertFalse(StringUtils.isNullOrWhitespace(token));
    }

    @Test
    public void getTimeZone() {
        String tz = new TimeZoneProvider().getTimeZone(0, 0);
        Log.d("TZTest", "tz = " + tz);
        Assert.assertFalse(StringUtils.isNullOrWhitespace(tz));
    }

    @Test
    public void serializationTest() throws WeatherException, IOException {
        WeatherProviderImpl provider = WeatherManager.getProvider(WeatherAPI.YAHOO);

        Stopwatch s = Stopwatch.createStarted();
        Weather weather = getWeather(provider);
        s.stop();
        Log.d("Serialzer", "JSON GetWeather Test: " + s.toString());

        for (int i = 0; i < 30; i++) {
            Stopwatch s2 = Stopwatch.createStarted();
            String json2 = JSONParser.serializer(weather, Weather.class);
            Weather desW2 = JSONParser.deserializer(json2, Weather.class);
            s2.stop();

            Log.d("Serialzer", "JSON2 Test " + (i + 1) + ": " + s2.toString());
        }
    }

    @Test
    public void getSunriseSetTime() throws WeatherException {
        final ZonedDateTime date = ZonedDateTime.now();
        String tz_long = "America/Los_Angeles";
        final LocationData locationData = new LocationData();
        locationData.setLatitude(47.6721646);
        locationData.setLongitude(-122.1706614);
        locationData.setTzLong(tz_long);
        Astronomy astro = AsyncTask.await(new CallableEx<Astronomy, WeatherException>() {
            @Override
            public Astronomy call() throws WeatherException {
                return new SolCalcAstroProvider().getAstronomyData(locationData, date);
            }
        });

        DateTimeFormatter fmt = DateTimeFormatter.ISO_LOCAL_DATE_TIME;
        Log.d("SolCalc", String.format(Locale.ROOT,
                "Sunrise: %s; Sunset: %s", astro.getSunrise().format(fmt), astro.getSunset().format(fmt)));
        Assert.assertTrue(astro.getSunrise() != LocalDateTime.MIN && astro.getSunset() != LocalDateTime.MIN);
    }

    @Test
    public void firebaseDBTest() throws ExecutionException, InterruptedException {
        long updateTime = Tasks.await(ImageDatabase.getLastUpdateTime());
        Assert.assertTrue(updateTime > 0);
    }
}

File: app/src/androidTest/java/com/thewizrd/simpleweather/UnitTests.kt
package com.thewizrd.simpleweather

import android.content.Context
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.location.Geocoder
import android.os.Bundle
import android.util.Log
import androidx.preference.PreferenceManager
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.google.common.base.Stopwatch
import com.google.common.collect.Iterables
import com.thewizrd.shared_resources.AppState
import com.thewizrd.shared_resources.ApplicationLib
import com.thewizrd.shared_resources.SimpleLibrary
import com.thewizrd.shared_resources.locationdata.LocationData
import com.thewizrd.shared_resources.locationdata.LocationProviderImpl
import com.thewizrd.shared_resources.locationdata.google.GoogleLocationProvider
import com.thewizrd.shared_resources.locationdata.weatherapi.WeatherApiLocationProvider
import com.thewizrd.shared_resources.tzdb.TimeZoneProvider
import com.thewizrd.shared_resources.utils.*
import com.thewizrd.shared_resources.utils.here.HEREOAuthUtils
import com.thewizrd.shared_resources.weatherdata.*
import com.thewizrd.shared_resources.weatherdata.aqicn.AQICNProvider
import com.thewizrd.shared_resources.weatherdata.model.Weather
import com.thewizrd.shared_resources.weatherdata.nws.SolCalcAstroProvider
import com.thewizrd.shared_resources.weatherdata.nws.alerts.NWSAlertProvider
import com.thewizrd.shared_resources.weatherdata.smc.SunMoonCalcProvider
import com.thewizrd.simpleweather.images.ImageDatabase
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.io.IOException
import java.time.LocalDateTime
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.*
import java.util.concurrent.ExecutionException

@RunWith(AndroidJUnit4::class)
class UnitTests {
    private lateinit var context: Context

    @Before
    fun init() {
        // Context of the app under test.
        context = ApplicationProvider.getApplicationContext()

        val app = object : ApplicationLib {
            override fun getAppContext(): Context {
                return context.applicationContext
            }

            override fun getPreferences(): SharedPreferences {
                return PreferenceManager.getDefaultSharedPreferences(appContext)
            }

            override fun registerAppSharedPreferenceListener() {}
            override fun unregisterAppSharedPreferenceListener() {}
            override fun registerAppSharedPreferenceListener(listener: OnSharedPreferenceChangeListener) {}
            override fun unregisterAppSharedPreferenceListener(listener: OnSharedPreferenceChangeListener) {}

            override fun getAppState(): AppState {
                return AppState.BACKGROUND
            }

            override fun isPhone(): Boolean {
                return true
            }

            override fun getProperties(): Bundle {
                return Bundle()
            }

            override fun getSettingsManager(): SettingsManager {
                return SettingsManager(appContext.applicationContext)
            }
        }

        // Needs to be called on main thread
        runBlocking(Dispatchers.Main.immediate) {
            SimpleLibrary.initialize(app)
        }

        // Start logger
        Logger.init(app.appContext)
        app.settingsManager.loadIfNeededSync()
    }

    @Throws(WeatherException::class)
    private suspend fun getWeather(
        providerImpl: WeatherProviderImpl,
        coordinate: Coordinate = Coordinate(
            47.6721646,
            -122.1706614
        ) /* Redmond, WA */
    ) = withContext(Dispatchers.IO) {
        val location = providerImpl.getLocation(coordinate)
        val locData = LocationData(location!!)
        return@withContext providerImpl.getWeather(locData)
    }

    @Throws(WeatherException::class)
    @Test
    fun getHEREWeather() {
        runBlocking(Dispatchers.Default) {
            val provider = WeatherManager.getProvider(WeatherAPI.HERE)
            val weather = getWeather(provider)
            Assert.assertTrue(weather.isValid)
        }
    }

    @Throws(WeatherException::class)
    @Test
    fun getMetNoWeather() {
        runBlocking(Dispatchers.Default) {
            val provider = WeatherManager.getProvider(WeatherAPI.METNO)
            val weather = getWeather(provider)
            Assert.assertTrue(weather.isValid)
        }
    }

    @Throws(WeatherException::class)
    @Test
    fun getNWSAlerts() {
        runBlocking(Dispatchers.Default) {
            val location = WeatherManager.getProvider(WeatherAPI.OPENWEATHERMAP)
                    .getLocation(Coordinate(47.6721646, -122.1706614))
            val locData = LocationData(location!!)
            val alerts = withContext(Dispatchers.IO) {
                NWSAlertProvider().getAlerts(locData)
            }
            Assert.assertNotNull(alerts)
        }
    }

    @Throws(WeatherException::class)
    @Test
    fun getNWSWeather() {
        runBlocking(Dispatchers.Default) {
            val provider = WeatherManager.getProvider(WeatherAPI.NWS)
            val weather = getWeather(provider)
            Assert.assertTrue(weather.forecast?.isNotEmpty() == true && weather.hrForecast?.isNotEmpty() == true)
            Assert.assertTrue(weather.isValid)
        }
    }

    @Throws(WeatherException::class)
    @Test
    fun getOWMWeather() {
        runBlocking(Dispatchers.Default) {
            val provider = WeatherManager.getProvider(WeatherAPI.OPENWEATHERMAP)
            val weather = getWeather(provider)
            Assert.assertTrue(weather.isValid)
        }
    }

    @Throws(WeatherException::class)
    @Test
    fun getWUnlockedWeather() {
        runBlocking(Dispatchers.Default) {
            val provider = WeatherManager.getProvider(WeatherAPI.WEATHERUNLOCKED)
            val weather = getWeather(provider)
            Assert.assertTrue(!weather.forecast.isNullOrEmpty() && !weather.hrForecast.isNullOrEmpty())
            Assert.assertTrue(weather.isValid)
        }
    }

    @Test
    fun getHEREOAuthToken() {
        runBlocking(Dispatchers.Default) {
            val token = withContext(Dispatchers.IO) {
                HEREOAuthUtils.getBearerToken(true)
            }
            Assert.assertFalse(token.isNullOrEmpty())
        }
    }

    @Test
    fun getTimeZone() {
        runBlocking(Dispatchers.Default) {
            val tz = withContext(Dispatchers.IO) {
                TimeZoneProvider().getTimeZone(0.0, 0.0)
            }
            Log.d("TZTest", "tz = $tz")
            Assert.assertFalse(tz.isNullOrEmpty())
        }
    }

    @Throws(WeatherException::class)
    @Test
    fun getAQIData() {
        runBlocking(Dispatchers.Default) {
            val locationData = LocationData().apply {
                latitude = 47.6721646
                longitude = -122.1706614
                tzLong = "America/Los_Angeles"
            }

            val aqi = withContext(Dispatchers.IO) {
                AQICNProvider().getAirQualityData(locationData)
            }
            Assert.assertNotNull(aqi)
        }
    }

    @Test
    @Throws(WeatherException::class, IOException::class)
    fun serializationTest() {
        runBlocking(Dispatchers.Default) {
            val provider = WeatherManager.getProvider(WeatherAPI.NWS)

            val s = Stopwatch.createStarted()
            val weather = getWeather(provider)
            s.stop()
            Log.d("Serialzer", "JSON GetWeather Test: $s")

            for (i in 0..29) {
                val s2 = Stopwatch.createStarted()
                val json2 = withContext(Dispatchers.Default) {
                    JSONParser.serializer(weather, Weather::class.java)
                }
                val desW2 = withContext(Dispatchers.Default) {
                    JSONParser.deserializer(json2, Weather::class.java)
                }
                s2.stop()

                Log.d("Serialzer", "JSON2 Test " + (i + 1) + ": " + s2.toString())
            }
        }
    }

    @Throws(WeatherException::class)
    @Test
    fun getSunriseSetTime() {
        runBlocking(Dispatchers.Default) {
            val date = ZonedDateTime.now()

            val locationData = LocationData().apply {
                latitude = 47.6721646
                longitude = -122.1706614
                tzLong = "America/Los_Angeles"
            }

            val astro = withContext(Dispatchers.Default) {
                SolCalcAstroProvider().getAstronomyData(locationData, date)
            }

            val fmt = DateTimeFormatter.ISO_LOCAL_DATE_TIME
            Log.d("SolCalc", String.format(Locale.ROOT,
                    "Sunrise: %s; Sunset: %s", astro.sunrise.format(fmt), astro.sunset.format(fmt)))
            Assert.assertTrue(astro.sunrise !== LocalDateTime.MIN && astro.sunset !== LocalDateTime.MIN)
        }
    }

    @Test
    @Throws(ExecutionException::class, InterruptedException::class)
    fun firebaseDBTest() {
        runBlocking(Dispatchers.Default) {
            val updateTime = ImageDatabase.getLastUpdateTime()
            Assert.assertTrue(updateTime > 0)
        }
    }

    @Test
    @Throws(Exception::class)
    fun simpleAstroTest() {
        runBlocking(Dispatchers.Default) {
            val date = ZonedDateTime.now()
            val locationData = LocationData().apply {
                latitude = 47.6721646
                longitude = -122.1706614
                tzLong = "America/Los_Angeles"
            }
            val astro = withContext(Dispatchers.Default) {
                SunMoonCalcProvider().getAstronomyData(locationData, date)
            }

            val fmt = DateTimeFormatter.ISO_LOCAL_DATE_TIME
            Log.d("SMC", String.format(Locale.ROOT,
                    "Sunrise: %s; Sunset: %s, Moonrise: %s; Moonset: %s", astro.sunrise.format(fmt), astro.sunset.format(fmt), astro.moonrise.format(fmt), astro.moonset.format(fmt)))
            if (astro.moonPhase != null) {
                Log.d("SMC", String.format(Locale.ROOT,
                        "Moonphase: %s", astro.moonPhase.phase.name))
            }

            Assert.assertTrue(astro.sunrise !== LocalDateTime.MIN && astro.sunset !== LocalDateTime.MIN && astro.moonrise !== LocalDateTime.MIN && astro.moonset !== LocalDateTime.MIN)
        }
    }

    @Test
    @Throws(IOException::class)
    fun androidAutoCompleteLocTest() {
        runBlocking(Dispatchers.Default) {
            Assert.assertTrue(Geocoder.isPresent())
            val geocoder = Geocoder(context, Locale.getDefault())
            val addressList = withContext(Dispatchers.IO) {
                geocoder.getFromLocationName("Redmond", 5) // Redmond
            }
            Assert.assertFalse(addressList.isNullOrEmpty())
        }
    }

    @Test
    @Throws(IOException::class)
    fun androidGeocoderTest() {
        runBlocking(Dispatchers.Default) {
            Assert.assertTrue(Geocoder.isPresent())
            val geocoder = Geocoder(context, Locale.getDefault())
            //List<Address> addressList = geocoder.getFromLocation(47.6721646, -122.1706614, 1); // Washington
            val addressList = withContext(Dispatchers.IO) {
                geocoder.getFromLocation(51.5073884, -0.1334347, 1) // London
            }
            Assert.assertFalse(addressList.isNullOrEmpty())
            val result = addressList!![0]
            Assert.assertNotNull(result)
        }
    }

    @Test
    @Throws(WeatherException::class)
    fun placesAPITest() {
        runBlocking(Dispatchers.Default) {
            val locationProvider = GoogleLocationProvider()
            val locations = withContext(Dispatchers.IO) {
                locationProvider.getLocations("Berlin, Germany", null)
            }
            Assert.assertFalse(locations.isNullOrEmpty())

            val queryVM = Iterables.getFirst(locations, null)
            Assert.assertNotNull(queryVM)

            val idModel = locationProvider.getLocationFromID(queryVM!!)
            /*
            LocationQueryViewModel nameModel = locationProvider.getLocationFromName(queryVM);
             */
            Assert.assertNotNull(idModel)
            //Assert.assertNotNull(nameModel);
        }
    }

    @Test
    @Throws(WeatherException::class)
    fun weatherAPILocationTest() {
        runBlocking(Dispatchers.Default) {
            val locationProvider: LocationProviderImpl = WeatherApiLocationProvider()
            val locations = withContext(Dispatchers.IO) {
                locationProvider.getLocations("Redmond, WA", null)
            }
            Assert.assertFalse(locations.isNullOrEmpty())

            val queryVM = locations.find { it.locationName.startsWith("Redmond, ") }
            Assert.assertNotNull(queryVM)

            val nameModel = locationProvider.getLocationFromName(queryVM!!)
            Assert.assertNotNull(nameModel)
        }
    }

    @Throws(WeatherException::class)
    @Test
    fun getMeteoFranceWeather() {
        runBlocking(Dispatchers.Default) {
            val provider = WeatherManager.getProvider(WeatherAPI.METEOFRANCE)
            val weather = getWeather(provider, Coordinate(48.85, 2.34))
            Assert.assertTrue(weather.isValid)
        }
    }
}